management-api-agent-hcd-cc4/src/main/java/org/apache/cassandra/transport/UnixSocketServerHcd.java

```diff
@@ -284,18 +319,64 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> ou
 
             promise = new VoidChannelPromise(ctx.channel(), false);
 
-            Message.Response response =
-                Dispatcher.processRequest(
-                    (ServerConnection) connection, startup, ClientResourceLimits.Overload.NONE);
-
-            if (response.type.equals(Message.Type.AUTHENTICATE))
-              // bypass authentication
-              response = new ReadyMessage();
-
-            outbound = response.encode(inbound.header.version);
-            ctx.writeAndFlush(outbound, promise);
-            logger.debug("Configured pipeline: {}", ctx.pipeline());
-            break;
+            // More Converged Cassandra/Core 4 changes for Async processing. This is generally a
+            // copy of upstream's InitConnectionHandler.
+
+            // Try to get the newer processInit static method
+            try {
+              Method processInit =
+                  Dispatcher.class.getDeclaredMethod(
+                      "processInit", ServerConnection.class, StartupMessage.class);
+              ((CompletableFuture<Message.Response>)
+                      processInit.invoke(null, (ServerConnection) connection, startup))
+                  .whenComplete(
+                      (response, error) -> {
+                        if (error == null) {
+                          Envelope encoded = response.encode(inbound.header.version);
+                          ctx.writeAndFlush(encoded, promise);
+                          logger.debug("Configured pipeline: {}", ctx.pipeline());
+                        } else {
+                          ErrorMessage message =
+                              ErrorMessage.fromException(
+                                  new ProtocolException(
+                                      String.format("Unexpected error %s", error.getMessage())));
+                          Envelope encoded = message.encode(inbound.header.version);
+                          ctx.writeAndFlush(encoded);
+                        }
+                      });
+              break;
+            } catch (NoSuchMethodException nsme) {
+              // try the older processRequest method
+              try {
+                Method processRequest =
+                    Dispatcher.class.getDeclaredMethod(
+                        "processRequest",
+                        ServerConnection.class,
+                        StartupMessage.class,
+                        ClientResourceLimits.Overload.class);
+                Message.Response response =
+                    (Message.Response)
+                        processRequest.invoke(
+                            null,
+                            (ServerConnection) connection,
+                            startup,
+                            ClientResourceLimits.Overload.NONE);
+                if (response.type.equals(Message.Type.AUTHENTICATE))
```

Comment:
For safety, this should be with { } instead of relying on indentation and a single line response. We don't use this type of "old bastard C coding who like bugs" in management-api elsewhere.

Btw, in this PR there are 3 ways of writing the same code, which feels weird. First the normal, then one with shorter if else conditional and then this one without {}.

Reply:
Good catch. I'll clean it up

Reply:
Actually, I'm glad you brought this up because I missed the same "if message type is AUTHENTICATE" in the code block above.